# LGBM: HyperOpt of The Parameters

Inspired by https://www.kaggle.com/prashant111/lightgbm-classifier-in-python

In [1]:
import pandas as pd
import numpy as np

import datetime as dt
from typing import Tuple, List, Dict

from sklearn.metrics import auc, roc_curve


from sklearn.model_selection import train_test_split, KFold, GridSearchCV, cross_validate
from sklearn.preprocessing import StandardScaler, LabelEncoder

import lightgbm as lgbm
from lightgbm import LGBMClassifier

from hyperopt import STATUS_OK, Trials, fmin, hp, tpe
from hyperopt.pyll.base import scope

from functools import partial

import warnings
warnings.filterwarnings('ignore')

C:\ProgramData\Anaconda3\lib\site-packages\numpy\_distributor_init.py:32: UserWarning: loaded more than 1 DLL from .libs:
C:\ProgramData\Anaconda3\lib\site-packages\numpy\.libs\libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll
C:\ProgramData\Anaconda3\lib\site-packages\numpy\.libs\libopenblas.NOIJJG62EMASZI6NYURL6JBKM4EVBGM7.gfortran-win_amd64.dll
  stacklevel=1)


In [2]:
# main flow
start_time = dt.datetime.now()
print("Started at ", start_time)

Started at  2021-03-23 21:17:38.889791


In [3]:
# read data
in_kaggle = False


def get_data_file_path(is_in_kaggle: bool) -> Tuple[str, str, str]:
    train_path = ''
    test_path = ''
    sample_submission_path = ''

    if is_in_kaggle:
        # running in Kaggle, inside the competition
        train_path = '../input/tabular-playground-series-mar-2021/train.csv'
        test_path = '../input/tabular-playground-series-mar-2021/test.csv'
        sample_submission_path = '../input/tabular-playground-series-mar-2021/sample_submission.csv'
    else:
        # running locally
        train_path = 'data/train.csv'
        test_path = 'data/test.csv'
        sample_submission_path = 'data/sample_submission.csv'

    return train_path, test_path, sample_submission_path


In [4]:
%%time
# get the training set and labels
train_set_path, test_set_path, sample_subm_path = get_data_file_path(in_kaggle)

train = pd.read_csv(train_set_path)
test = pd.read_csv(test_set_path)
target = train.target

test_id = test.id

subm = pd.read_csv(sample_subm_path)

Wall time: 4.59 s


In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 32 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   id      300000 non-null  int64  
 1   cat0    300000 non-null  object 
 2   cat1    300000 non-null  object 
 3   cat2    300000 non-null  object 
 4   cat3    300000 non-null  object 
 5   cat4    300000 non-null  object 
 6   cat5    300000 non-null  object 
 7   cat6    300000 non-null  object 
 8   cat7    300000 non-null  object 
 9   cat8    300000 non-null  object 
 10  cat9    300000 non-null  object 
 11  cat10   300000 non-null  object 
 12  cat11   300000 non-null  object 
 13  cat12   300000 non-null  object 
 14  cat13   300000 non-null  object 
 15  cat14   300000 non-null  object 
 16  cat15   300000 non-null  object 
 17  cat16   300000 non-null  object 
 18  cat17   300000 non-null  object 
 19  cat18   300000 non-null  object 
 20  cont0   300000 non-null  float64
 21  cont1   30

In [6]:
cols_to_drop = ['id', 'cat5', 'cat7', 'cat8', 'cat10']

cat_cols = [feature for feature in train.columns if 'cat' in feature and feature not in cols_to_drop ]
cont_cols = [feature for feature in train.columns if 'cont' in feature and feature not in cols_to_drop]

all_features = cat_cols + cont_cols

def preprocess(df, encoder=None,
               scaler=None, cols_to_drop=None,
               cols_to_encode=None, cols_to_scale=None):
    """
    Preprocess input data
    :param df: DataFrame with data
    :param encoder: encoder object with fit_transform method
    :param scaler: scaler object with fit_transform method
    :param cols_to_drop: columns to be removed
    :param cols_to_encode: columns to be encoded
    :param cols_to_scale: columns to be scaled
    :return: DataFrame
    """

    if encoder:
        for col in cols_to_encode:
            df[col] = encoder.fit_transform(df[col])

    if scaler:
        for col in cols_to_scale:
            df[col] = scaler.fit_transform(df[col].values.reshape(-1, 1))

    if cols_to_drop:
        df = df.drop(cols_to_drop, axis=1)

    return df

In [7]:
train = preprocess(train, encoder=LabelEncoder(), scaler=StandardScaler(),
                  cols_to_drop=cols_to_drop, cols_to_encode=cat_cols,
                  cols_to_scale=cont_cols)

# encoder=LabelEncoder()
test = preprocess(test, encoder=LabelEncoder(), scaler=StandardScaler(),
                 cols_to_drop=cols_to_drop, cols_to_encode=cat_cols,
                 cols_to_scale=cont_cols)

In [8]:
# ------------------------------------------------------------------------------
# Parameters
# ------------------------------------------------------------------------------
N_FOLDS = 10
N_ESTIMATORS = 30000
SEED = 2021
BAGGING_SEED = 48

In [9]:
lgb_search_params = {
        "num_leaves": scope.int(hp.quniform("num_leaves", 31, 250, 1)),
        "max_depth": scope.int(hp.quniform("max_depth", 10, 24, 1)),
        'learning_rate': hp.uniform('learning_rate', 0.01, 0.3),
        'subsample_for_bin': scope.int(hp.quniform('subsample_for_bin', 10000, 300000, 10000)),
        'min_data_in_leaf': scope.int(hp.qloguniform('min_data_in_leaf', 0, 6, 1)),
        'min_split_gain': hp.uniform('min_split_gain', 0, 1.0),
        'min_child_samples': scope.int(hp.quniform("min_child_samples", 2, 700, 1)),
        "subsample": hp.uniform("subsample", 0.2, 1.0), #alias: bagging_fraction
        "colsample_bytree": hp.uniform("colsample_bytree", 0.5, 1.0), # alias: feature_fraction
        'reg_alpha': hp.uniform('reg_alpha', 1e-5, 1.0),
        'reg_lambda': hp.uniform('reg_lambda', 0, 50),
        'min_sum_hessian_in_leaf': None, #overrides min_child_weight
        'min_child_weight': hp.loguniform('min_child_weight', -16, 5), #also aliases to min_sum_hessian
        'min_child_samples': None, #overridden by min_data_in_leaf
        'n_estimators': 2000
}

#         'boosting_type': hp.choice('boosting_type', [{'boosting_type': 'gbdt'}, 
#                                                 {'boosting_type': 'dart'},
#                                                 {'boosting_type': 'goss'}]),
#
#    'boosting_type_choice': hp.choice('boosting_type', 
#                                      [{'boosting_type': 'gbdt', 'subsample': hp.uniform('gbdt_subsample', 0.5, 1)}, 
#                                       {'boosting_type': 'dart', 'subsample': hp.uniform('dart_subsample', 0.5, 1)},
#                                       {'boosting_type': 'goss', 'subsample': 1.0}]),
#        'n_jobs': -1,

def lgbm_search(params):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=22)

    model = LGBMClassifier(boosting_type='gbdt', **params) # boosting_type='gbdt',

    evaluation = [(X_test, y_test)]

    model.fit(X_train, y_train,
              eval_set=evaluation, eval_metric='auc', categorical_feature=cat_cols,
              early_stopping_rounds=100, verbose=False)

    val_preds = model.predict(X_test)
    
    fpr, tpr, thresholds = roc_curve(y_test, val_preds, pos_label = 1)
    auc_score = auc(fpr, tpr)
    
    neg_auc_score = -1 * auc_score

    return {"loss": neg_auc_score, "status": STATUS_OK}

In [10]:
# search for model

X = train[all_features]
y = target


trials = Trials()

best_hyperparams = fmin(fn=lgbm_search,
                       space=lgb_search_params,
                       algo=tpe.suggest,
                       max_evals=100,
                       trials=trials)

100%|█████████████████████████████████████████████| 100/100 [34:18<00:00, 20.59s/trial, best loss: -0.7762041348019278]


In [11]:
best_hyperparams

{'colsample_bytree': 0.9263854460240344,
 'learning_rate': 0.10129388430065381,
 'max_depth': 21.0,
 'min_child_weight': 2.1767634493243653e-06,
 'min_data_in_leaf': 19.0,
 'min_split_gain': 0.3769322564423919,
 'num_leaves': 140.0,
 'reg_alpha': 0.22869561967261584,
 'reg_lambda': 40.80465168202993,
 'subsample': 0.5880585727954809,
 'subsample_for_bin': 140000.0}

In [13]:
# count each class label observations to calculate the class label weights
df = train['target'].value_counts()
df

0    220539
1     79461
Name: target, dtype: int64

In [21]:
# setting class weights as inspired by 
# https://datascience.stackexchange.com/questions/54043/differences-between-class-weight-and-scale-pos-weight-in-lightgbm
number_of_positive_samples = df[1]
number_of_negative_samples = df[0]

class_weight = {0: (number_of_positive_samples / number_of_negative_samples), 
                1: (number_of_negative_samples / number_of_positive_samples)}


lgbm_params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'random_state': SEED,
    'n_estimators': 30000,
    'n_jobs': -1,
    'cat_feature': [x for x in range(len(cat_cols))],
    'bagging_seed': SEED,
    'feature_fraction_seed': SEED,
    'class_weight': class_weight,
    
    
    'colsample_bytree': 0.9263854460240344,
    'learning_rate': 0.006752926, # 0.10129388430065381,
    'max_depth': 21,
    'min_child_weight': 2.1767634493243653e-06,
    'min_data_in_leaf': 19,
    'min_split_gain': 0.3769322564423919,
    'num_leaves': 140,
    'reg_alpha': 0.22869561967261584,
    'reg_lambda': 40.80465168202993,
    'subsample': 0.5880585727954809,
    'subsample_for_bin': 140000
}


In [22]:
class_weight

{0: 0.3603036197679322, 1: 2.775437006833541}

In [23]:
#X = X.abs()
y = train["target"]
kf = KFold(n_splits=N_FOLDS, shuffle=True, random_state=1)
oof_2 = np.zeros(len(train))
score_list = []
fold = 1
test_preds_2 = []

# view accuracy
from sklearn.metrics import accuracy_score


for train_index, test_index in kf.split(train):
    X_train, X_val = train.iloc[train_index], train.iloc[test_index]
    y_train, y_val = y.iloc[train_index], y.iloc[test_index]

    
    dtrain = X_train[all_features]
    dvalid = X_val[all_features]
    
    model = LGBMClassifier(**lgbm_params) 

    evaluation = [(dvalid, y_val)]

    model.fit(dtrain, y_train,
              eval_set=evaluation, eval_metric='auc', categorical_feature=cat_cols,
              early_stopping_rounds=100, verbose=False)

    val_preds = model.predict(dvalid)
    
    fpr, tpr, thresholds = roc_curve(y_val, val_preds, pos_label = 1)
    auc_score = auc(fpr, tpr)


    print(f"AUC Score Fold-{fold} : {auc_score}")

    oof_2[test_index] = val_preds

    
    score_list.append(auc_score)
    test_preds_2.append(model.predict_proba(test[all_features])[:, 1])
    fold+=1
    

np.mean(score_list)

AUC Score Fold-1 : 0.7831048302914687
AUC Score Fold-2 : 0.7838855269897482
AUC Score Fold-3 : 0.7854728740056997
AUC Score Fold-4 : 0.7828964445038262
AUC Score Fold-5 : 0.7822928790573617
AUC Score Fold-6 : 0.7811681812169002
AUC Score Fold-7 : 0.7844727067165023
AUC Score Fold-8 : 0.7847252074341798
AUC Score Fold-9 : 0.7833282305118042
AUC Score Fold-10 : 0.7831199220465526


0.7834466802774044

In [24]:
print(np.mean(score_list))
score_list

0.7834466802774044


[0.7831048302914687,
 0.7838855269897482,
 0.7854728740056997,
 0.7828964445038262,
 0.7822928790573617,
 0.7811681812169002,
 0.7844727067165023,
 0.7847252074341798,
 0.7833282305118042,
 0.7831199220465526]

In [25]:
train["2_preds"] = oof_2
test["2_preds"] = np.mean(test_preds_2,axis=0)

fpr, tpr, thresholds = roc_curve(train["target"], train["2_preds"], pos_label = 1)
auc_score = auc(fpr, tpr)
    
print(f"AUC Score Final : {auc_score}")

AUC Score Final : 0.7834404637957391


In [26]:
# save prediction
subm_df = test[["2_preds"]]
subm_df['id'] = test_id
subm_df = subm_df.rename(columns={"2_preds": "target"})
subm_df.to_csv("hyperopt_single_lgb_model.csv",index=False)

In [27]:
print('We are done. That is all, folks!')
finish_time = dt.datetime.now()
print("Finished at ", finish_time)
elapsed = finish_time - start_time
print("Elapsed time: ", elapsed)

We are done. That is all, folks!
Finished at  2021-03-24 00:02:56.731379
Elapsed time:  2:45:17.841588


# Historical Record Tracking

## Baseline, with all features

- Final validation AUC locally - 0.7776592687653098
- public LB 0.882171

```
lgbm_params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'random_state': SEED,
    'n_estimators': N_ESTIMATORS,
    'n_jobs': -1,
    'cat_feature': [x for x in range(len(cat_cols))],
    'bagging_seed': SEED,
    'feature_fraction_seed': SEED,
    
    
    'colsample_bytree': 0.8785258997000931,
    'learning_rate': 0.014051289,
    'max_depth': 11,
    'min_child_weight': 5.5803474120805336e-06,
    'min_data_in_leaf': 85,
    'min_split_gain': 0.23035992374930991,
    'num_leaves': 248,
    'reg_alpha': 0.5030438494317617,
    'reg_lambda': 27.412351627541167,
    'subsample': 0.5725588226322971,
    'subsample_for_bin': 160000
}
```

## Single lgbm without non-important cat features

- AUC Score Final (locally) : 0.7860753330981496
- public lb 0.88821

cat cols dropped: ['cat5', 'cat7', 'cat8', 'cat10']

(jump from top 66% to top 47% in public lb as of Mar 6, 2021)

```
lgbm_params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'random_state': SEED,
    'n_estimators': N_ESTIMATORS,
    'n_jobs': -1,
    'cat_feature': [x for x in range(len(cat_cols))],
    'bagging_seed': SEED,
    'feature_fraction_seed': SEED,
    
    
    'colsample_bytree': 0.7671990402625448,
     'learning_rate': 0.019767284,
     'max_depth': 15,
     'min_child_weight': 0.0002777396522195953,
     'min_data_in_leaf': 57,
     'min_split_gain': 0.21136452616481655,
     'num_leaves': 175,
     'reg_alpha': 0.11063887315135824,
     'reg_lambda': 18.001114258982668,
     'subsample': 0.9818248529067929,
     'subsample_for_bin': 180000
}


```

## Single lgbm without non-important cat features, with target class label weights

- AUC Score Final (locally) : 0.784074298578011
- public lb 0.88821

cat cols dropped: ['cat5', 'cat7', 'cat8', 'cat10']


```
lgbm_params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'random_state': SEED,
    'n_estimators': 30000,
    'n_jobs': -1,
    'cat_feature': [x for x in range(len(cat_cols))],
    'bagging_seed': SEED,
    'feature_fraction_seed': SEED,
    'class_weight': class_weight,
    
    
    'colsample_bytree': 0.7671990402625448,
     'learning_rate': 0.019767284,
     'max_depth': 15,
     'min_child_weight': 0.0002777396522195953,
     'min_data_in_leaf': 57,
     'min_split_gain': 0.21136452616481655,
     'num_leaves': 175,
     'reg_alpha': 0.11063887315135824,
     'reg_lambda': 18.001114258982668,
     'subsample': 0.9818248529067929,
     'subsample_for_bin': 180000
}

class_weight = {0: 0.3603036197679322, 1: 2.775437006833541}

```

## Single lgbm without non-important cat features, with target class label weights, smaller n_rounds

- AUC Score Final (locally) : 0.7862662928595938
- public lb 0.88821 (slightly worse then n_rounds=30000 

cat cols dropped: ['cat5', 'cat7', 'cat8', 'cat10']


```
lgbm_params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'random_state': SEED,
    'n_estimators': 15000,
    'n_jobs': -1,
    'cat_feature': [x for x in range(len(cat_cols))],
    'bagging_seed': SEED,
    'feature_fraction_seed': SEED,
    'class_weight': class_weight,
    
    
    'colsample_bytree': 0.7671990402625448,
     'learning_rate': 0.039534568,
     'max_depth': 15,
     'min_child_weight': 0.0002777396522195953,
     'min_data_in_leaf': 57,
     'min_split_gain': 0.21136452616481655,
     'num_leaves': 175,
     'reg_alpha': 0.11063887315135824,
     'reg_lambda': 18.001114258982668,
     'subsample': 0.9818248529067929,
     'subsample_for_bin': 180000
}

class_weight = {0: 0.3603036197679322, 1: 2.775437006833541}

```

##  Single lightgbm, without noisy cat features, with class weights, tuned with hyperopt on the negative auc loss 

- AUC Score Final (locally) : 0.7834404637957391
- public lb ... 

cat cols dropped: ['cat5', 'cat7', 'cat8', 'cat10']

```
lgbm_params = {
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'random_state': SEED,
    'n_estimators': 30000,
    'n_jobs': -1,
    'cat_feature': [x for x in range(len(cat_cols))],
    'bagging_seed': SEED,
    'feature_fraction_seed': SEED,
    'class_weight': class_weight,
    
    
    'colsample_bytree': 0.9263854460240344,
    'learning_rate': 0.006752926, # 0.10129388430065381,
    'max_depth': 21,
    'min_child_weight': 2.1767634493243653e-06,
    'min_data_in_leaf': 19,
    'min_split_gain': 0.3769322564423919,
    'num_leaves': 140,
    'reg_alpha': 0.22869561967261584,
    'reg_lambda': 40.80465168202993,
    'subsample': 0.5880585727954809,
    'subsample_for_bin': 140000
}

class_weight = {0: 0.3603036197679322, 1: 2.775437006833541}

```